Do not assume all pages have changed

pathawks opened the pull request:
`<lastmod>` [is an optional tag](http://www.sitemaps.org/protocol.html#lastmoddef).

Including it as we currently do has the advantage of letting search engines know if nothing on the site has changed for quite a while (as in, the site has not been regenerated), but the disadvantage of telling search engines that every page on the site has changed each time the site is regenerated, even if no pages have been updated.

I'm not really sure which is worse&hellip;

erkki commented:
Why not keep post.time there if it's present? Agreed on not lying but wondering if there's better ways to provide real information here.

pathawks commented:
Where would `post.time` come from?

erkki commented:
yeah figured it must come from somewhere, for posts it seems post.date is a convention, for pages difficult indeed without making it a configurable field

erkki commented:
wonder if it's in any way feasible to hook into the actual mtimes assuming identical content files don't get bumped on generation

pathawks commented:
> wonder if it's in any way feasible to hook into the actual mtimes assuming identical content files don't get bumped on generation

The trouble with this approach is that you'd have to check - not just the modified time of the source file - but all layouts and includes used to generate the page. If the layout has a newer modified time than the page's source file, which one would you use?

I like the idea. I wish there were an easy way to provide a modified time for everything. I'm just not sure how it could/should be done.

erkki commented:
if jekyll would actually compare the final rendered/generated result before overwriting it'd be easier, haven't checked whether that's the case or not (might be a sep pull to jekyll if not)

pathawks commented:
See also jekyll/jekyll#2260

Does [jekyll-last-modified-at](https://github.com/gjtorikian/jekyll-last-modified-at) work for pages as well as posts? That'd solve the whole problem right there. @gjtorikian?

pathawks commented:
Turns out, yes it does!

If [jekyll-last-modified-at](https://github.com/gjtorikian/jekyll-last-modified-at) is installed, we can use it to figure out a last modified date for pages the same way we use it for posts!

erkki commented:
sounds promising :)

pathawks commented:
I have updated the pull request to only add `<lastmod>` if we actually know what the modified date is through a plugin like [jekyll-last-modified-at](https://github.com/gjtorikian/jekyll-last-modified-at)

parkr commented:
We should have two test cases to test whether it's included if last_modified_at is present.

pathawks commented:
Two test cases? One for pages and one for collections?

Did something happen to Travis? My green checkmarks are missing.

pathawks commented:
*Sometimes* (but not always) it seems that all the posts are set to the same mtime with **jekyll-last-modified-at**.

https://travis-ci.org/jekyll/jekyll-sitemap/jobs/47089163#L210-L220

There is probably something I am doing wrong. I've not written many tests like this before.
*Any* guidance would be much appreciated.

parkr commented:
First, rebase on current master. That'll be a better starting point. :)

pathawks commented:
This will also certainly conflict with #33.
Any chance of merging that one first, before I rebase everything?

parkr commented:
This one can't be merged.

pathawks commented:
**jekyll-last-modified-at** can either be turned on for a whole site, or not included for a whole site.

If we want to test how this plugin works with **jekyll-last-modified-at**, we also definitely want to continue testing how it works without it.

Given this, should I create a *second* test site that includes the **jekyll-last-modified-at** gem and tests that `<lastmod>` is output in a sane way? Or should I hack around this by adding `last_modified_at` to the front matter of a couple pages and testing that *that* is output correctly in `<lastmod>`?

wisq commented:
Hm, I'm getting weird behaviour here.  I took the latest `jekyll-sitemap` master, forked it, applied this branch, and ended up with https://github.com/wisq/jekyll-sitemap/commit/888a65535df218297438bdaf8422b21faee7279e

If I include it directly from Github via my `Gemfile`:

```Gemfile
gem 'jekyll-last-modified-at'
gem 'jekyll-sitemap', :github => 'wisq/jekyll-sitemap', :ref => '888a65535df218297438bdaf8422b21faee7279e'
```

… `jekyll-last-modified-at` doesn't seem to be able to retrieve any modified dates.  In my sitemap, all my posts are dated as their official publishing date (not their asset changed date), and all my pages have **no** date (as opposed to stock behaviour **without** this PR, which would be giving them the current date).

However, if I include it via a `:path`:

```Gemfile
gem 'jekyll-last-modified-at'
gem 'jekyll-sitemap', :path => '../jekyll-sitemap'
```

… it works as I would expect: Posts **and** pages both have the date of their most recent asset change.

My best guess is that somehow, including it via `git` is somehow messing with `last-modified-at`s git file modification time sensing.  I don't know why or how, though, since `../jekyll-sitemap` is a git repo too, so I would expect it to exhibit the same behaviour.

Anyway, this may not be an issue with this PR; it's probably just some bug in `jekyll-last-modified-at`.  Figured I'd mention it, though, because it's downright *weird*.

wisq commented:
Okay, actually, it seems to randomly either work (for ALL pages) or not work.  The gemfile thing was a red herring. :/

Continuing to investigate.

wisq commented:
Okay, yeah, it's what I thought.  Sometimes the SiteMap generator runs before the LastModifiedAt generator, and everything is broken if it does.

See https://github.com/wisq/jekyll-sitemap/commit/12ba411290f666dc886338239600b12ed159300b for a possible fix.

pathawks commented:
> Sometimes the SiteMap generator runs before the LastModifiedAt generator, and everything is broken if it does.

:collision:

parkr commented:
The priority of `last_modified_at` should be raised. I think the priority here is already set to `:lowest`.

wisq commented:
I don't believe the priority was already `:lowest`, because then my change would've *raised* the priority and made the problem *always* occur — yet (in my several dozen test runs) it seemed to make the problem *never* occur instead.

I'm pretty sure they both default to `:normal` priority.  That said, logically, yes, `last-modified-at` should probably be increased so that other `:normal` priority plugins can make use of it.

One more point to make about this: This will break any sites that attempt to sitemap dynamic pages that don't have an equivalent page in the source tree.  For example, `jekyll-tagging` generates dynamic pages based on a layout.

This was never a problem before this PR because the sitemap didn't care about the source files, but with this PR in place, combining `jekyll-sitemap`, `jekyll-last-modified-at`, and `jekyll-tagging` will raise an error when it generates a tag page.

I ended up solving this locally using a combination of https://github.com/wisq/jekyll-tagging/commit/068eb44ed7ab64f9552d104c1eca3863526ec5ae and https://github.com/wisq/jekyll-sitemap/commit/2429b2cdc31de455da7694c6a4e84ba8e0398c12 .  Of course, this assumes that I don't want those pages in the sitemap; maybe some people do, and they should just be marked with the current date, or the most recent last-modified-at of all the pages they index.

pathawks commented:
If `jekyll-tagging` just creates an index of links to other pages, then those *probably* don't belong in a sitemap anyway.

wisq commented:
Yeah, that's reasonable.  Still need a way to omit them, though.  As far as I know, there's nothing in a `Page` object that immediately says "this didn't come from a real file".

parkr commented:
Any other changes to this PR necessary?

pathawks commented:
> Any other changes to this PR necessary?

Yes. Need to write tests.
Should I create a second test site specifically for testing with `last-modified-at`?

> Still need a way to omit them
Could `jekyll-tagging` be modified to set `page.sitemap` false? Or is this something more general that we need to tackle here?

wisq commented:
Yeah, setting sitemap to false for tag pages would be the quick solution here.  Only concern would be that I don't know how many other plugins might also generate files from scratch and break here.

I wonder if we can just use a tag/filter that tries `page.last_modified_at` but rescues `Errno::ENOENT`.

pathawks commented:
> I wonder if we can just use a tag/filter that tries page.last_modified_at but rescues Errno::ENOENT.

This is probably something that should be fixed inside [jekyll-last-modified-at](https://github.com/gjtorikian/jekyll-last-modified-at)?

wisq commented:
Well, it's maybe(?) reasonable to freak out if you try to get the last-modified-at of a file that can't logically have one, rather than just lying about it / returning nil.  Not sure.

parkr commented:
> Should I create a second test site specifically for testing with last-modified-at?

eah.

pathawks commented:
I'm sorry, I really have no idea how to do that.

parkr commented:
> I'm sorry, I really have no idea how to do that.

That's ok! You'd want to setup a different gemfile and setup the travis matrix to have one set of tests which run with the gemfile, and one set of tests which run without. That's a naïve approach, but it'd work.

pathawks commented:
If I run:

```sh
bundle exec rspec spec/test_jekyll-last-modified-at.rb
```

My new test file will run, and my tests will execute.

How do I make this happen automatically?

pathawks commented:
Ok, I think I've got it worked out.

I added a new test file which is _not_ executed with the default **rspec** tests. This is important because it `require 'jekyll-last-modified-at'` which would screw everything else up that doesn't expect it.

Then, I added a line to script/cibuild to execute the new test file, separately from everything else.

I am currently not checking for a specific time with static files. The others all use the time of commit, but static files seem to use the file system modified time, which will always be changing on Travis.
gjtorikian/jekyll-last-modified-at#33

***

Did I do this right? Does this appear satisfactory? Ready for one last rebase?

pathawks commented:
> I am currently not checking for a specific time with static files.

This is not supported by *jekyll-last-modified-at* anyway; best to just continue to use the system mtime, I guess.

Otherwise, everything is all rebased and ready to rock. I hope.

pathawks commented:
Rebased.

parkr commented:
Cool, thanks! @gjtorikian, this adds in *optional* support to jekyll-last-modified-at (i.e if it's there, great, otherwise ignore it). Do you think it'd fly in GitHub land?

pathawks commented:
FWIW, this has been available for posts since [v0.6.0](https://github.com/jekyll/jekyll-sitemap/releases/tag/v0.6.0).
This PR just extends support to pages and docs, as well as adding a couple tests.

gjtorikian commented:
> Do you think it'd fly in GitHub land?

That's cool with me. Unfortunately we can't rely on `last-modified-at` working on Pages any time soonish, because of the way it deals with builds. But locally this should still be fine, no?
